lib/kitchen/element_enumerator_factory.rb

```diff
@@ -65,38 +61,34 @@ def or_with(other_factory)
       )
     end
 
-    def self.apply_default_css_or_xpath_and_normalize(css_or_xpath: nil, default_css_or_xpath: nil)
-      [css_or_xpath || '$'].flatten.map do |item|
-        item.gsub(/\$/, [default_css_or_xpath].flatten.join(', '))
-      end
-    end
-
     protected
 
-    def build_within_element(element, css_or_xpath:)
-      enumerator_class.new(css_or_xpath: css_or_xpath) do |block|
+    def build_within_element(element, search_query:)
+      enumerator_class.new(search_query: search_query) do |block|
         grand_ancestors = element.ancestors
         parent_ancestor = Ancestor.new(element)
 
-        num_sub_elements = 0
-
-        element.raw.search(*css_or_xpath).each_with_index do |sub_node, index|
+        element.raw.search(*search_query.css_or_xpath).each_with_index do |sub_node, index|
           sub_element = ElementFactory.build_from_node(
             node: sub_node,
             document: element.document,
             element_class: sub_element_class,
-            default_short_type: Utils.search_path_to_type(css_or_xpath),
+            default_short_type: search_query.as_type,
             detect_element_class: detect_sub_element_class
           )
 
-          # If the provided `css_or_xpath` has already been counted, we need to uncount
+          next unless search_query.conditions_match?(sub_element)
```

Comment:
This is the meat of the PR.  Filters out elements based on the `only` and `except` conditions in `search_query`